Fruit shop solution

VladGural opened the pull request.

kovalskiy7777 commented on src/main/java/core/basesyntax/Main.java:
```diff
@@ -0,0 +1,46 @@
+
+public class Main {
+    public static void main(String[] args) {
+        String sourceFilePath = "src/main/resources/12-09-2021.csv";
+        String resultFilePath = "src/main/resources/12-09-2021result.csv";
```
```suggestion
```
Эти переменные можно сделать константами

kovalskiy7777 commented on src/main/java/core/basesyntax/Main.java:
```diff
@@ -0,0 +1,46 @@
+        String sourceFilePath = "src/main/resources/12-09-2021.csv";
+        String resultFilePath = "src/main/resources/12-09-2021result.csv";
+        Map<String, ActivityHandler> activities = new HashMap<>();
+        activities.put("b", new BalanceActivityHandler());
+        activities.put("s", new SupplyActivityHandler());
+        activities.put("p", new PurchaseActivityHandler());
+        activities.put("r", new ReturnActivityHandler());
```
```suggestion
```
Для символов которые означают тип операции лучше воспользоваться енамом

kovalskiy7777 commented on src/main/java/core/basesyntax/files/ReaderFile.java:
```diff
@@ -0,0 +1,25 @@
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class ReaderFile implements Reader {
```
```suggestion
public class ReaderImpl implements Reader {
```
И у Богдана в лекциях был нейминг подобно этому. Когда имплементируешь интерфейс то добавляешь к имени просто Impl. Возможно это не принципиально

VladGural replied:
Как я понял Reader это абстракция и я из нее реализую класс ReaderFile именно File, просто тут же можно реализовать например ReaderDB  для чрения из БД и уже его передавать ReaderServise

VladGural replied:
fixed

kovalskiy7777 commented on src/main/java/core/basesyntax/servise/ReaderServiceImpl.java:
```diff
@@ -0,0 +1,35 @@
+        List<List<String>> resultList = new ArrayList<>();
+        List<String> lines = reader.read();
+        for (int i = 0; i < lines.size(); i++) {
+            List<String> parsedLine = Arrays.asList(lines.get(i).strip().split(","));
```
```suggestion

```
(",") - можно сделать константой

kovalskiy7777 commented on src/main/java/core/basesyntax/servise/WriterServiceImpl.java:
```diff
@@ -0,0 +1,24 @@
+    @Override
+    public void writeData(Map<String, Integer> map) {
+        List<String> resultList = new ArrayList<>();
+        resultList.add("fruit,quantity");
```
```suggestion

```
это также константа

kovalskiy7777 commented on src/main/java/core/basesyntax/servise/WriterServiceImpl.java:
```diff
@@ -0,0 +1,24 @@
+        List<String> resultList = new ArrayList<>();
+        resultList.add("fruit,quantity");
+        for (Map.Entry<String, Integer> entry : map.entrySet()) {
+            resultList.add(entry.getKey() + "," + entry.getValue());
```
```suggestion
```
и здесь

kovalskiy7777 commented on src/main/resources/12-09-2021.csv:
```diff
@@ -0,0 +1,9 @@
+    r,apple,10
+    p,apple,20
+    p,banana,5
+    s,banana,50
```
Мне кажется, что нужно еще добавить файл с отчетом. 
И не уверен, что это принципиально, но у тебя нет ни единой модельки. Например, ты бы мог сделать фрукт отдельным классом, или сделать класс, который бы  олицетворял то, что у тебя приходит в файле

Iryna-U commented on src/main/java/core/basesyntax/Main.java:
```diff
@@ -0,0 +1,64 @@
+        shopService.availableFruitsAfterWorkShift();
+    }
+
+    enum Activities {
+        BALANCE("b"),
+        SUPPLY("s"),
+        PURCHASE("p"),
+        RETURN("r");
```
Цей єнам краще винести в окремим клас. Або ж він може бути у класі FruitRecordDto про який написала нижче

VladGural commented:
added in name's implementation class prefix Impl
created model FruitRecordDto to transfer data
added result file to repository
removed enum Activities from main class to package lib

Iryna-U commented on src/main/java/core/basesyntax/servise/activity/PurchaseActivityHandlerImpl.java:
```diff
@@ -0,0 +1,17 @@
+    public void act(String fruit, String quantity, Map<String, Integer> goalMap) {
+        Integer intQuantity = Integer.valueOf(quantity);
+        if (goalMap.get(fruit) == null) {
+            throw new RuntimeException("Balance isn't first activity");
```
тут ми саме цей месседж маємо виводити?

VladGural replied:
Ну я имел в виду, что перед любыми операциями, в том числе и Purchase, должен вначале идти Balance, наверное надо написать Balance wasn't first activity

Iryna-U commented on src/main/java/core/basesyntax/servise/ShopServiceImpl.java:
```diff
@@ -0,0 +1,31 @@
+        parsedLines = readerService.readData();
+        for (FruitRecordDto line : parsedLines) {
+            activityStrategy.getActivity(line.getTypeOperation())
+                    .act(line.getFruit(), line.getQuantity(), goalMap);
```
goalMap - репрезинтує в нас базу даних з фруктами. Давай краще створимо окремий клас Storage,  де буде зберігітись ця мапа, наприклад з назвою - fruitsDataBase. Тоді в метод .act() тобі не доведеться кожен раз передавати мапу
```
public class Storage {
    public static final Map<Fruit, Integer> fruitsDataBase = new HashMap<>();
}
```

VladGural commented:
changed RuntimeException message
added class storage
